### DISCUSSION TITLE
yandex_music.exceptions.BadRequest: invalid_grant login or password is not valid

### DISCUSSION THREAD

**Discussion by dec04** [MarshalX/yandex-music-api#439] [Q&A] [answered]
**Опишите ошибку**
При вводе корректных данных постоянно падает ошибка invalid_grant login or password is not valid

**Воспроизведение**
Шаги для воспроизведения бага:
1. Войти под своим аккаунтом используя метов Client.from_credentials()

**Ожидаемое поведение**
Корректный вход

**Логи**
Traceback (most recent call last):
  File "<input>", line 2, in <module>
  File "/root/PycharmProjects/yandex-music-api/venv/lib/python3.8/site-packages/yandex_music/client.py", line 206, in from_credentials
    token = cls(*args, **kwargs).generate_token_by_username_and_password(
  File "/root/PycharmProjects/yandex-music-api/venv/lib/python3.8/site-packages/yandex_music/client.py", line 72, in wrapper
    result = method(*args, **kwargs)
  File "/root/PycharmProjects/yandex-music-api/venv/lib/python3.8/site-packages/yandex_music/client.py", line 279, in generate_token_by_username_and_password
    result = self._request.post(url, data, timeout=timeout, *args, **kwargs)
  File "/root/PycharmProjects/yandex-music-api/venv/lib/python3.8/site-packages/yandex_music/utils/request.py", line 264, in post
    result = self._request_wrapper(
  File "/root/PycharmProjects/yandex-music-api/venv/lib/python3.8/site-packages/yandex_music/utils/request.py", line 215, in _request_wrapper
    raise BadRequest(message)
yandex_music.exceptions.BadRequest: invalid_grant login or password is not valid

**Дополнительная информация**
Поля корректные, сутки все работало отлично, я даже написал бот для discord. Но теперь не работает. Попытка использовать капчу завершилась неудачей.

**Comment by MarshalX** [comment, marked as answer]
@dec04 Яндекс переезжают на новую систему авторизации. Все кто хоть раз зашёл с новой авторизации теперь не имеют возможности входить старым способом. Протестировать новую авторизацию можно на dev ветке библиотеки. Подробнее тут: #414

**Comment by dec04** [comment]
> @dec04 Яндекс переезжают на новую систему авторизации. Все кто хоть раз зашёл с новой авторизации теперь не имеют возможности входить старым способом. Протестировать новую авторизацию можно на dev ветке библиотеки. Подробнее тут: #414

А есть примеры использования новой системы авторизации?

**Comment by MarshalX** [comment]
@dec04 Методы остались прежними. Чутка изменилась обработка капчи и порядок аргументов для капчи

**Comment by dec04** [comment]
> @dec04 Методы остались прежними. Чутка изменилась обработка капчи и порядок аргументов для капчи

Технически - нужно просто переехать в новую ветку и в теории должно работать?

**Comment by MarshalX** [comment]
@dec04 Да, при условии, если Вы не писали обработку капчи при множестве неправильных попыток входа)

**Comment by dec04** [comment]
Я не писал еще обработку капчи, но попробовал перейти на новую ветку, с теми же данными для credentials

```python
from yandex_music import Client 


Client.from_credentials("******@yandex.ru", "*******PWD.")
```

```python
Traceback (most recent call last):
  File "<input>", line 2, in <module>
  File "/root/PycharmProjects/yandex-music-api/venv/lib/python3.8/site-packages/yandex_music/client.py", line 219, in from_credentials
    x_token = client._send_authentication_password(
  File "/root/PycharmProjects/yandex-music-api/venv/lib/python3.8/site-packages/yandex_music/client.py", line 72, in wrapper
    result = method(*args, **kwargs)
  File "/root/PycharmProjects/yandex-music-api/venv/lib/python3.8/site-packages/yandex_music/client.py", line 344, in _send_authentication_password
    raise BadRequest(result)
yandex_music.exceptions.BadRequest: {'status': 'error', 'state': 'auth_challenge', 'errors': ['action.required_external_or_native']}

```

**Comment by MarshalX** [comment]
@dec04 Сталкивались с такой проблемой. Сложная штука. Вот реф где почитать про это: 

"action.required_external_or_native": "Аккаунт не поддерживается ([подробнее](https://github.com/AlexxIT/YandexStation/issues/103))"

**Comment by dec04** [comment]
> @dec04 Сталкивались с такой проблемой. Сложная штука. Вот реф где почитать про это:
> 
> "action.required_external_or_native": "Аккаунт не поддерживается ([подробнее](https://github.com/AlexxIT/YandexStation/issues/103))"

Спасибо за реф, судя по описанию проблема заключалась в том что у маня какой-то "неправильный" аккаунт. Как описано в ссылке, у меня аккаунт времен "Народа", но это не так. Скорее всего это связано с тем что мой аккаунт использовался как рабочий, когда я работал Агентом/Пешеходом. Включение двухфакторки - помогло.

Последовательность действий которая заработала для моего аккаунта:
 1. Включение двухфакторной авторизации;
 2. Пройти все шаги для смены двухфакторной авторизации;
 3. Выйти, зайти;
 4. Выключить двухфакторку, сменить пароль на старый;

**Comment by MarshalX** [comment]
@dec04 Спасибо за маленькое исследование и расписанные шаги!
